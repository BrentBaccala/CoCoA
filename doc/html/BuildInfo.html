<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML>
<HEAD>
<META NAME="generator" CONTENT="http://txt2tags.sf.net">
<LINK REL="stylesheet" TYPE="text/css" HREF="cocoalib-doc.css">
<TITLE>      BuildInfo</TITLE>
</HEAD><BODY BGCOLOR="white" TEXT="black">
<P ALIGN="center"><CENTER><H1>      BuildInfo</H1>
<FONT SIZE="4">
<I>      &copy;  2005,2007 John Abbott</I><BR>
      GNU Free Documentation License, Version 1.2
</FONT></CENTER>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>
    <UL>
    <LI><A HREF="#toc1">Examples</A>
    <LI><A HREF="#toc2">User documentation</A>
    <LI><A HREF="#toc3">Maintainer documentation</A>
    <LI><A HREF="#toc4">Bugs, Shortcomings and other ideas</A>
    </UL>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>
<P>
      <center><a href="index.html">CoCoALib Documentation Index</a></center>
</P>
<A NAME="toc1"></A>
<H2>Examples</H2>
<UL>
<LI><A HREF="../../examples/index.html#ex-BuildInfo.C">ex-BuildInfo.C</A>
<LI><A HREF="../../examples/index.html#ex-limits.C">ex-limits.C</A>
</UL>

<P></P>
<A NAME="toc2"></A>
<H2>User documentation</H2>
<P>
The constant in <CODE>BuildInfo</CODE> allows you to find out which version of
CoCoALib you are using.  The function <CODE>BuildInfo::PrintAll</CODE> prints out all
the build information on the <CODE>ostream</CODE> passed in as argument -- you should
include this information whenever you report a bug.
</P>
<P>
There is one string constant which contains the version number of the
library: it is of the form <CODE>A.bcde</CODE> where <CODE>A</CODE> is the major
version, <CODE>bc</CODE> is the minor version, and <CODE>de</CODE> is the patch level.
Note that there are always precisely 4 digits after the point (even if
they are all zero).
</P>
<UL>
<LI><CODE>BuildInfo::version</CODE> -- a C string containing the CoCoALib version number.
</UL>

<P></P>
<P>
NOTE: if you happen upon a copy of libcocoa.a and want to find out which
version it is, you can use the following Unix/Linux command:
</P>
<PRE>
    strings libcocoa.a | egrep "CoCoA::BuildInfo"
</PRE>
<P></P>
<P>
This should print out three lines informing you of the library version,
the compiler used, and the compiler flags used when creating libcocoa.a.
</P>
<A NAME="toc3"></A>
<H2>Maintainer documentation</H2>
<P>
I chose to put the constants and function in their own namespace to
emphasise that they go together.
</P>
<P>
There are actually four string constants, but only one is supposed to
be publicly accessible (because I cannot imagine why anyone would want
access to the other three).  I made the constants C strings because it
seemed simpler than using C++ strings.  The three constants
<CODE>VersionMesg</CODE>, <CODE>CompilerMesg</CODE>, and <CODE>CompilerFlagsMesg</CODE> contain
the substring <CODE>CoCoA::BuildInfo</CODE> so that the egrep trick described
above will produce the version/compiler information directly.
</P>
<P>
I made <CODE>BuildInfo::PrintAll</CODE> leave a blank line before and after its
message so that it would stand out better from other output produced
by the program.
</P>
<A NAME="toc4"></A>
<H2>Bugs, Shortcomings and other ideas</H2>
<P>
The constants are not C++ strings -- is this really a bug?
</P>
<P>
Should the three constants <CODE>VersionMesg</CODE>, <CODE>CompilerMesg</CODE>, and
[CompilerFlagsMesg] be hidden or public?  Until someone convinces me there
is a good reason to make them public, they'll stay private.
</P>

<!-- html code generated by txt2tags 2.5 (http://txt2tags.sf.net) -->
<!-- cmdline: txt2tags -t html -o html/BuildInfo.html -i txt/BuildInfo.txt -->
</BODY></HTML>
