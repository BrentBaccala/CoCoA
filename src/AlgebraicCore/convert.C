//   Copyright (c)  2007,2009  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.

#include "CoCoA/convert.H"
#include "CoCoA/BigRat.H"
#include "CoCoA/IntOperations.H"
#include "CoCoA/ring.H"

#include <limits>
using std::numeric_limits;
#include <cmath>
using std::ldexp;
using std::floor;

namespace CoCoA
{

  //-------------------------------------------------------
  // BigInt

  bool IsConvertible(long& lhs, const BigInt& src)
  {
    if (!mpz_fits_slong_p(mpzref(src))) return false;
    lhs = mpz_get_si(mpzref(src));
    return true;
  }

  bool IsConvertible(int& lhs, const BigInt& src)
  {
    if (!mpz_fits_sint_p(mpzref(src))) return false;
    lhs = mpz_get_si(mpzref(src));
    return true;
  }

  bool IsConvertible(unsigned long& lhs, const BigInt& src)
  {
    if (!mpz_fits_ulong_p(mpzref(src))) return false;
    lhs = mpz_get_ui(mpzref(src));
    return true;
  }

  bool IsConvertible(unsigned int& lhs, const BigInt& src)
  {
    if (!mpz_fits_uint_p(mpzref(src))) return false;
    lhs = mpz_get_ui(mpzref(src));
    return true;
  }


  //-------------------------------------------------------
  // BigRat

  bool IsConvertible(long& lhs, const BigRat& src)
  {
    // Simple rather than fast -- makes unnecessary copy of num(src)
    return IsOneDen(src) && IsConvertible(lhs, num(src));
  }

  bool IsConvertible(int& lhs, const BigRat& src)
  {
    // Simple rather than fast -- makes unnecessary copy of num(src)
    return IsOneDen(src) && IsConvertible(lhs, num(src));
  }

  bool IsConvertible(unsigned long& lhs, const BigRat& src)
  {
    // Simple rather than fast -- makes unnecessary copy of num(src)
    return IsOneDen(src) && IsConvertible(lhs, num(src));
  }

  bool IsConvertible(unsigned int& lhs, const BigRat& src)
  {
    // Simple rather than fast -- makes unnecessary copy of num(src)
    return IsOneDen(src) && IsConvertible(lhs, num(src));
  }


  //-------------------------------------------------------
  // double with long, BigInt, BigRat

  bool IsConvertible(long& n, double z)
  {
    //??? BUG should handle infinity and Nan???
    if (z != std::floor(z)) return false;
    if (z < numeric_limits<long>::min() || z > numeric_limits<long>::max()) return false;
    n = static_cast<long>(z); // already checked that z is integer and in range.
    return true;
  }


  // NB ALL sufficiently large doubles are considered to be integers
  bool IsConvertible(BigInt& N, double z)
  {
    //??? BUG should handle infinity and Nan???
    if (z != std::floor(z)) return false;
    mpz_set_d(mpzref(N), z);
    return true;
  }




  bool IsConvertible(BigRat& Q, double z)
  {
    //??? BUG should handle infinity and Nan???
    if (z == 0) { Q = 0; return true; }
    mpq_set_d(mpqref(Q), z);  // BUG Don't know what happens if this fails.
    return true;
  }


  // Sets z = N if possible.
  bool IsConvertible(double& z, const BigInt& N)
  {
    if (IsZero(N)) { z = 0.0; return true; }
    // I prefer to do this manually because mpz_get_d does not give
    // good guarantees about what happens when overflow occurs.
    long ExpN;
    double MantissaN = mpz_get_d_2exp(&ExpN, mpzref(N));
    if (ExpN > numeric_limits<double>::max_exponent)
    {
      return false;
    }
    z = ldexp(MantissaN, ExpN);
    return true;
  }


  // Sets z = num/den if possible.
  // Underflow sets z to 0 and returns true.
  // Overflow returns false (and does not change z).
  bool IsConvertible(double& z, const BigRat& Q)
  {
    if (IsZero(Q)) { z = 0.0; return true; }
    // I prefer to do this manually rather than use mpq_get_d because
    // num and den may not be coprime, and the cost of computing the
    // gcd could be high.  Also mpq_get_d does not give nice guarantees
    // about what happens in the case of over-/under-flow.
    long Nexp;
    double N = mpz_get_d_2exp(&Nexp,mpq_numref(mpqref(Q)));
    long Dexp;
    double D = mpz_get_d_2exp(&Dexp, mpq_denref(mpqref(Q)));
    long ExpAns = Nexp-Dexp;
    double MantissaAns = N/D;
    if (MantissaAns >= 1 || MantissaAns <= -1)
    { MantissaAns /= 2; ++ExpAns; }

    // I believe this is the right interpretation of the exponent limits...
    if (ExpAns < numeric_limits<double>::min_exponent)
    {
      // We have underflow, so convert to 0 and indicate "success"
      z = 0.0;
      return true;
    }

    if (ExpAns > numeric_limits<double>::max_exponent)
    {
      // We have overflow, so indicate "failure".
      return false;
    }

    z = ldexp(MantissaAns, ExpAns); // Cannot overflow or underflow.
    return true;
  }


  //-------------------------------------------------------
  // RingElem to long, BigInt, BigRat

  bool IsConvertible(long& n, ConstRefRingElem x)
  {
    BigInt N;
    return (IsConvertible(N, x) && IsConvertible(n, N));
  }

  bool IsConvertible(BigInt& N, ConstRefRingElem x)
  {
    return IsInteger(N, x);
  }

  bool IsConvertible(BigRat& N, ConstRefRingElem x)
  {
    return IsRational(N, x);
  }


} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/convert.C,v 1.23 2014/04/02 15:55:45 abbott Exp $
// $Log: convert.C,v $
// Revision 1.23  2014/04/02 15:55:45  abbott
// Summary: Reorganized file contents (more sensible order now, I hope)
// Author: JAA
//
// Revision 1.22  2013/03/26 14:56:06  abbott
// Updated the conversion fns (in ptic removed procedure "convert");
// numerous consequential changes.
//
// Revision 1.21  2013/02/15 16:32:44  abbott
// Removed IsInRange (moved to MachineInt).
// Added IsConvertible(BigInt, RingElem) and IsConvertible(BigRat, RingElem).
//
// Revision 1.20  2012/05/28 09:18:20  abbott
// Created IntOperations which gathers together all operations on
// integers (both big and small).  Many consequential changes.
//
// Revision 1.19  2011/11/09 15:35:00  bigatti
// -- moved include MachineInt into convert.H
//
// Revision 1.18  2011/11/09 14:29:37  bigatti
// -- renamed MachineInteger --> MachineInt
//
// Revision 1.17  2011/08/27 20:50:47  abbott
// Added include directive for MachineInt.
//
// Revision 1.16  2011/08/24 10:32:04  bigatti
// -- renamed QQ --> BigRat
// -- sorted #include
//
// Revision 1.15  2011/08/14 15:52:16  abbott
// Changed ZZ into BigInt (phase 1: just the library sources).
//
// Revision 1.14  2011/08/12 16:08:32  abbott
// Hastily added conversion fns for BigInt (like those for ZZ).
//
// Revision 1.13  2011/03/22 20:26:12  abbott
// Added  IsConvertible(long&, double).
// Activated commented out ConvertTo template.
// Checking in because some other files need the new code.
//
// Revision 1.12  2010/04/23 13:21:45  abbott
// Conversion from string to ZZ now defaults to decimal (and not the quirky C standard).
//
// Revision 1.11  2010/03/22 11:51:37  abbott
// Simplified convert fn from string to BigRat -- now uses GMP's implementation.
//
// Revision 1.10  2009/12/23 22:27:28  abbott
// Added conversions from BigRat to machine integers.
//
// Revision 1.9  2009/12/23 18:53:51  abbott
// Major change to conversion functions:
//   convert(..) is now a procedure instead of a function
//   IsConvertible(..) replaces the former convert(..) function
//   Added new NumericCast conversion function (placeholder for BOOST feature)
//   Consequent changes in code which uses these features.
//
// Revision 1.8  2009/12/11 11:46:32  abbott
// Changed fn  convert  into  IsConvertible.
// Added template procedure  convert.
// New version because change is not backward compatible.
//
// Revision 1.7  2009/07/08 12:26:53  abbott
// Added floor and ceil functions for BigRats.
// Added example program for BigRats.
// Minor correction to convert.C; minor cleaning to ex-ZZ1.C
//
// Revision 1.6  2009/07/02 16:32:11  abbott
// Consequential changes stemming from new class BigRat, and modified interface to the member
// function RingBase::myIsRational.  Also some new conversion functions.
//
// Revision 1.5  2008/10/07 15:45:22  abbott
// Changed ErrorInfo objects so they include the name of their own error ID.
// Changed catch statements to catch const objects.
// Removed calls to the member fn which accessed the error ID member of an
// ErrorInfo; now you simply compare directly with the error ID (makes the
// code easier to read).
//
// Revision 1.4  2007/10/30 17:14:07  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.3  2007/05/22 22:51:39  abbott
// Changed name of fn ndigits to NumDigits.
// Changed return type of exponent and NumDigits.
// Changed some exceptions from nonstandard to the appropriate standard one.
//
// Revision 1.2  2007/03/28 15:34:03  abbott
// Underflow during conversion to a double no longer causes "failure" but now
// converts the value to zero and yields "success".
//
// Revision 1.1  2007/03/23 18:38:42  abbott
// Separated the "convert" functions (and CheckedCast) into their own files.
// Many consequential changes.  Also corrected conversion to doubles.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.3  2007/03/03 15:24:39  cocoa
// Changed 2006 to 2007.
//
// Revision 1.2  2006/10/06 14:04:14  cocoa
// Corrected position of #ifndef in header files.
// Separated CoCoA_ASSERT into assert.H from config.H;
// many minor consequential changes (have to #include assert.H).
// A little tidying of #include directives (esp. in Max's code).
//
