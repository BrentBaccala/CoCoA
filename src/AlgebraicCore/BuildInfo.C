//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
