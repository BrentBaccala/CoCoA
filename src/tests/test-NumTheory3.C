//   Copyright (c)  2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/convert.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/NumTheory.H"
#include "CoCoA/error.H"


#include <iostream>
using std::cerr;
using std::cout;
using std::endl;

using namespace CoCoA;

// Handy macro for making assertions.
#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)

void program()
{
  // Test for functions related to continued fractions.
  GlobalManager CoCoAFoundations;

  // Generator/iterator for continued fraction "quotients"
  BigRat x(-101,100);
  while (x < 0)
  {
    ContFracIter cfi(x);
    cout << "The continued fraction quotients of " << x << " are:";
    while (!IsEnded(cfi))
      cout << "  " << *cfi++;
    cout << endl;
    x += BigRat(1,10);
  }

  // Make a rational from its cont frac quotients...
  ContFracApproximant approx;
  for (ContFracIter it(x); !IsEnded(it); ++it)
    approx.myAppendQuot(quot(it));
  TEST_ASSERT(approx.myRational() == x);

  // Compute the approximants for a rational...
  x = BigRat(-123,100);
  for (CFApproximantsIter it(x); !IsEnded(it); ++it)
    cout << "Approximant: " << *it << endl;

  // CFApprox...
  TEST_ASSERT(CFApprox(ConvertTo<BigRat>(1.4142135), BigRat(1,1)) == BigRat(1,1));
  TEST_ASSERT(CFApprox(ConvertTo<BigRat>(-1.4142135), BigRat(1,1)) == BigRat(-2,1));
  TEST_ASSERT(CFApprox(ConvertTo<BigRat>(1.4142135), ConvertTo<BigRat>(0.01)) == BigRat(17,12));
  TEST_ASSERT(CFApprox(ConvertTo<BigRat>(-1.4142135), ConvertTo<BigRat>(0.01)) == BigRat(-17,12));

  TEST_ASSERT(SimplestBigRatBetween(BigRat(1,1), BigRat(2,1)) == BigRat(1,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-2,1), BigRat(-1,1)) == BigRat(-1,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-1,1), BigRat(1,1)) == BigRat(0,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(0,1), BigRat(1,1)) == BigRat(0,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-1,1), BigRat(0,1)) == BigRat(0,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(1,1), BigRat(4,3)) == BigRat(1,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(4,3), BigRat(2,1)) == BigRat(2,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-4,3), BigRat(-1,1)) == BigRat(-1,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-2,1), BigRat(-4,3)) == BigRat(-2,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(4,3), BigRat(10,3)) == BigRat(2,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-10,3), BigRat(-4,3)) == BigRat(-2,1));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(123,100), BigRat(135,100)) == BigRat(4,3));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-135,100), BigRat(-123,100)) == BigRat(-4,3));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(5,2), BigRat(299,100)) == BigRat(5,2));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-299,100), BigRat(-5,2)) == BigRat(-5,2));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(4181, 2584), BigRat(6765,4181)) == BigRat(4181,2584));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-6765,4181), BigRat(-4181, 2584)) == BigRat(-4181,2584));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(225,157), BigRat(268,187)) == BigRat(225,157));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-268,187), BigRat(-225,157)) == BigRat(-225,157));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(1393,972), BigRat(268,187)) == BigRat(268,187));
  TEST_ASSERT(SimplestBigRatBetween(BigRat(-268,187), BigRat(-1393,972)) == BigRat(-268,187));
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
