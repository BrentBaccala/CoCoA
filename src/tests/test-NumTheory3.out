The continued fraction quotients of -101/100 are:  -2  1  99
The continued fraction quotients of -91/100 are:  -1  11  9
The continued fraction quotients of -81/100 are:  -1  5  3  1  4
The continued fraction quotients of -71/100 are:  -1  3  2  4  3
The continued fraction quotients of -61/100 are:  -1  2  1  1  3  2  2
The continued fraction quotients of -51/100 are:  -1  2  24  2
The continued fraction quotients of -41/100 are:  -1  1  1  2  3  1  1  2
The continued fraction quotients of -31/100 are:  -1  1  2  4  2  3
The continued fraction quotients of -21/100 are:  -1  1  3  1  3  5
The continued fraction quotients of -11/100 are:  -1  1  8  11
The continued fraction quotients of -1/100 are:  -1  1  99
Approximant: -2
Approximant: -1
Approximant: -5/4
Approximant: -11/9
Approximant: -16/13
Approximant: -123/100
